### PR TITLE
feat: added the precision and recall metrics for QA accuracy

### PR DESCRIPTION
**Pull request by bilalaws**
*Description of changes:* This pull request adds the Precision and Recall metrics for the Question Answering task. Previous metrics do not capture the cases where one of the target output or model output is short and the other one is long.

For instance, consider the question `Did RMS Titanic sink in 1912?` If the target output is `Yes` and the model output is `Yes. The ship indeed sank in 1912. It was the largest ship at the time <some long text>` then the existing metrics will give a low score even though the answer is correct. The recall metric added in this PR will be `1.0` indicating that all of the target output words are contained within the model output. The precision metric operates in the opposite direction and measures what fraction of words in the model output are found in the target output.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/fmeval/eval_algorithms/qa_accuracy.py [hunk 1]
```diff
@@ -116,6 +118,46 @@ def _f1_score(model_output: str, target_output: str, *, normalize_text: bool = F
...
+    Before computing precision, we normalize the text following the QuAC protocol.
+
+    :param model_output: The output of a model that we want to evaluate.
+    :param target_output: The reference or the "ground truth" output.
+    :param normalize_text: Normalize the text before computing f1.
+    :returns: Precision.
+    """
+    if normalize_text:  # pragma: no branch
+        model_output, target_output = (_normalize_text_quac_protocol(text) for text in (model_output, target_output))
+    ret = precision(reference=set(target_output.split(" ")), test=set(model_output.split(" ")))
```
Comment (franluca) [inline src/fmeval/eval_algorithms/qa_accuracy.py:134]: why set and not list? we want to discard repetitions of words?

File: src/fmeval/eval_algorithms/qa_accuracy.py [hunk 2]
```diff
@@ -116,6 +118,46 @@ def _f1_score(model_output: str, target_output: str, *, normalize_text: bool = F
...
+    Before computing recall, we normalize the text following the QuAC protocol.
+
+    :param model_output: The output of a model that we want to evaluate.
+    :param target_output: The reference or the "ground truth" output.
+    :param normalize_text: Normalize the text before computing f1.
+    :returns: Recall.
+    """
+    if normalize_text:  # pragma: no branch
+        model_output, target_output = (_normalize_text_quac_protocol(text) for text in (model_output, target_output))
+    ret = recall(reference=set(target_output.split(" ")), test=set(model_output.split(" ")))
```
Comment (franluca) [inline src/fmeval/eval_algorithms/qa_accuracy.py:154]: same comment

File: src/fmeval/eval_algorithms/qa_accuracy.py [hunk 3]
```diff
@@ -116,6 +118,46 @@ def _f1_score(model_output: str, target_output: str, *, normalize_text: bool = F
         return float(ret)
 
 
+def _precision(model_output: str, target_output: str, *, normalize_text: bool = False) -> float:
+    """
+    Given the model output and the target output, compute the precision.
+    Precision is the fraction of words in the prediction that are also found in the target output.
+    Before computing precision, we normalize the text following the QuAC protocol.
```
Comment (franluca) [inline src/fmeval/eval_algorithms/qa_accuracy.py:125]: what's the QuAC protocol? Can you add a link?

File: src/fmeval/eval_algorithms/qa_accuracy.py [hunk 4]
```diff
@@ -116,6 +118,46 @@ def _f1_score(model_output: str, target_output: str, *, normalize_text: bool = F
...
+
+    :param model_output: The output of a model that we want to evaluate.
+    :param target_output: The reference or the "ground truth" output.
+    :param normalize_text: Normalize the text before computing f1.
```
Comment (franluca) [inline src/fmeval/eval_algorithms/qa_accuracy.py:129]: .. before computing the precision

File: src/fmeval/eval_algorithms/qa_accuracy.py [hunk 5]
```diff
@@ -116,6 +118,46 @@ def _f1_score(model_output: str, target_output: str, *, normalize_text: bool = F
...
+
+def _recall(model_output: str, target_output: str, *, normalize_text: bool = False) -> float:
+    """
+    Given the model output and the target output, compute the recall.
+    Recall is the fraction of words in the target output that are also found in the answer.
+    Before computing recall, we normalize the text following the QuAC protocol.
+
+    :param model_output: The output of a model that we want to evaluate.
+    :param target_output: The reference or the "ground truth" output.
+    :param normalize_text: Normalize the text before computing f1.
```
Comment (franluca) [inline src/fmeval/eval_algorithms/qa_accuracy.py:149]: before computing the recall

File: src/fmeval/eval_algorithms/qa_accuracy.py (hunk 1 above)
Comment (polaschwoebel) [reply src/fmeval/eval_algorithms/qa_accuracy.py:134]: Valid question, but set seems standard. At least that's what the [NLTK metric](https://www.nltk.org/_modules/nltk/metrics/scores.html) assumes that we use here.

**Review by polaschwoebel — APPROVED** [review]

File: src/fmeval/eval_algorithms/qa_accuracy.py (hunk 3 above)
Comment (bilalaws) [reply src/fmeval/eval_algorithms/qa_accuracy.py:125]: The function `_normalize_text_quac_protocol` in the same file [here](https://github.com/aws/fmeval/blob/30cc7c4d2692f19e7fe7927f2f60bcde9da96ccd/src/fmeval/eval_algorithms/qa_accuracy.py#L78) has the info.

File: src/fmeval/eval_algorithms/qa_accuracy.py (hunk 5 above)
Comment (bilalaws) [reply src/fmeval/eval_algorithms/qa_accuracy.py]: fixed in the new commit

File: src/fmeval/eval_algorithms/qa_accuracy.py (hunk 4 above)
Comment (bilalaws) [reply src/fmeval/eval_algorithms/qa_accuracy.py]: fixed

**Review by polaschwoebel — APPROVED** [review]

**Review by franluca — APPROVED** [review]

**Review by malhotra18 — APPROVED** [review]

**Review by malhotra18 — APPROVED** [review]

**Review by polaschwoebel — APPROVED** [review]

**Pull request merged**